drivers/flash/flash_mcux_flexspi_hyperflash.c

```diff
@@ -0,0 +1,622 @@
+		LOG_ERR("failed to CFI");
+		return ret;
+	}
+
+	transfer.deviceAddress = (0x10 * 2);
+	transfer.port = config->port;
+	transfer.cmdType = kFLEXSPI_Read;
+	transfer.SeqNumber = 1;
+	transfer.seqIndex = READ_DATA;
+	transfer.data = buffer;
+	transfer.dataSize = 8;
+
+	ret = memc_flexspi_transfer(data->controller, &transfer);
+	if (ret != 0) {
+		LOG_ERR("failed to read id");
+		return ret;
+	}
+	buffer[1] &= 0xFFFF;
+	/* Check that the data read out is  unicode "QRY" in big-endian order */
+	if ((buffer[0] != 0x52005100) || (buffer[1] != 0x5900)) {
+		LOG_ERR("data read out is wrong!");
+		return -EINVAL;
+	}
+
+	writebuf[1] = 0xF0;
+	transfer.deviceAddress = 0;
+	transfer.port = config->port;
+	transfer.cmdType = kFLEXSPI_Write;
+	transfer.SeqNumber = 1;
+	transfer.seqIndex = WRITE_DATA;
+	transfer.data = (uint32_t *)writebuf;
+	transfer.dataSize = 2;
+
+	ret = memc_flexspi_transfer(data->controller, &transfer);
+	if (ret != 0) {
+		LOG_ERR("failed to exit");
+		return ret;
+	}
+
+	memc_flexspi_reset(data->controller);
+
+	return ret;
+}
+
+static int flash_flexspi_hyperflash_page_program(const struct device *dev, off_t
+		offset, const void *buffer, size_t len)
+{
+	const struct flash_flexspi_hyperflash_config *config = dev->config;
+	struct flash_flexspi_hyperflash_data *data = dev->data;
+	flexspi_transfer_t transfer = {
+		.deviceAddress = offset,
+		.port = config->port,
+		.cmdType = kFLEXSPI_Write,
+		.SeqNumber = 2,
+		.seqIndex = PAGE_PROGRAM,
+		.data = (uint32_t *)buffer,
+		.dataSize = len,
+	};
+
+	LOG_DBG("Page programming %d bytes to 0x%08x", len, offset);
+
+	return memc_flexspi_transfer(data->controller, &transfer);
+}
+
+static int flash_flexspi_hyperflash_read(const struct device *dev, off_t offset,
+		void *buffer, size_t len)
+{
+	const struct flash_flexspi_hyperflash_config *config = dev->config;
+	struct flash_flexspi_hyperflash_data *data = dev->data;
+	uint8_t *src = memc_flexspi_get_ahb_address(data->controller,
+			config->port,
+			offset);
+	if (!src) {
+		LOG_ERR("invalid src?");
+		return -EINVAL;
+	}
+
+	memcpy(buffer, src, len);
+
+	return 0;
+}
+
+static int flash_flexspi_hyperflash_write(const struct device *dev, off_t offset,
+		const void *buffer, size_t len)
+{
+	const struct flash_flexspi_hyperflash_config *config = dev->config;
+	struct flash_flexspi_hyperflash_data *data = dev->data;
+	size_t size = len;
+	uint8_t *src = (uint8_t *)buffer;
+	unsigned int key = 0;
+	int i;
+	int ret;
+
+	uint8_t *dst = memc_flexspi_get_ahb_address(data->controller,
+			config->port,
+			offset);
+	if (!dst) {
+		LOG_ERR("ahb invalid?");
+	}
+
+	if (memc_flexspi_is_running_xip(data->controller)) {
+		key = irq_lock();
+	}
+
+	while (len) {
+		/* If the offset isn't a multiple of the NOR page size, we first need
+		 * to write the remaining part that fits, otherwise the write could
+		 * be wrapped around within the same page
+		 */
+		i = MIN(SPI_HYPERFLASH_PAGE_SIZE - (offset %
+					SPI_HYPERFLASH_PAGE_SIZE), len);
```

Comment:
The SDK does not permit unaligned programming. Would this be better:
```
if ((offset % SPI_HYPERFLASH_PAGE_SIZE) != 0) {
    return -EINVAL;
}
```

Reply:
The settings subsystem with NVS backend uses unaligned offsets, not sure if it can be forced to uses full pages, which wouldn't very efficient though.

Reply:
You currently set `write_block_size` to `HYPERFLASH_WRITE_SIZE=1` which would allow the unaligned access.

Reply:
I'll have a look at it. Is there a sample or a test I can run to confirm it is working properly?

Reply:
I think you can use `samples/subsys/settings`

Reply:
I cannot get the sample to work with NVS. Something about "too many pages" so it returns -ENOMEM.
Any ideas?

Reply:
@Kludentwo, I think this is caused by the flash_area subsystem. You should be able to trick it by defining your hyperflash to have less than 255 "sectors/pages/...", just define only e.g. halve of your flash as the size. I don't think nvs needs unaligned access, it is required to set the write-block-size correct: that would be 16 for hyperflash (half page in hyperflash terminology). And it will then of course waste some space to keep the alignment requirement.

Reply:
@Laczen where did you find the write-block-size value (16)?